forem/forem
# [15 min fix] Fix some autoloading deprecation warnings

## What type of PR is this? (check all applicable)


## Description

When a console is load from scratch (`spring stop; bundle exec rails console`), Rails outputs quite a long deprecation message having to do with [autoloading](https://guides.rubyonrails.org/autoloading_and_reloading_constants.html) and the new autloader that was added in Rails 6.

Addressing those deprecations, with the help of the [official guide on autoloading](https://guides.rubyonrails.org/autoloading_and_reloading_constants.html) and [discourse's middlewares initializer](https://github.com/discourse/discourse/blob/dc6b547ed89f652b5406489d76140b76cf8e0d1d/config/initializers/200-first_middlewares.rb) removed some deprecations.

Unfortunately, I couldn't figure out how to deal with `ForemStatsClient` which is an odd pre-initialized constant we have lying around. 

I tried to use [Rails.application.reloader.to_prepare](https://guides.rubyonrails.org/autoloading_and_reloading_constants.html#autoloading-when-the-application-boots) to have the autoloader "digest" it correctly like we do in other cases *but*

https://github.com/forem/forem/blob/ff244a31eff9d8ad270dd7b0aa4c6ed5eac03223/app/lib/forem_stats_drivers/datadog_driver.rb#L17

has to access `config.middleware` which is already frozen when `.to_prepare` is called. The other thing I tried was to move the initialization inside `app/lib` but then it's not eager loaded correctly in production mode.

I wonder if there's something to be done to improve the code and make Rails happy 😕 

## QA Instructions, Screenshots, Recordings

You can run:

```
RAILS_ENV=production bundle exec rails middleware
```

to make sure `Middlewares::SetCookieDomain` is in the list.

You can run:

```
RAILS_ENV=test bundle exec rails middleware
```

to make sure `Middlewares::SetTimeZone` is in the list.

An additional check is to load the console in production mode and make sure there are no ddtrace related warnings: `RAILS_ENV=production bundle exec rails console`


Regarding deprecations, we went from:

```shell
Unset ENV variable: IMGPROXY_KEY.
Unset ENV variable: IMGPROXY_SALT.
W, [2021-05-18T18:08:12.386802 #47598]  WARN -- ddtrace: [ddtrace] Unable to patch Datadog::Contrib::RestClient::Integration (Available?: true, Loaded? false, Compatible? true, Patchable? false)
DEPRECATION WARNING: Initialization autoloaded the constants RateLimitChecker, ForemStatsDriver, ForemStatsDrivers, ActsAsForemStatsDriver, and ForemStatsDrivers::DatadogDriver.

Being able to do this is deprecated. Autoloading during initialization is going
to be an error condition in future versions of Rails.

Reloading does not reboot the application, and therefore code executed during
initialization does not run again. So, if you reload RateLimitChecker, for example,
the expected changes won't be reflected in that stale Class object.

These autoloaded constants have been unloaded.

In order to autoload safely at boot time, please wrap your code in a reloader
callback this way:

    Rails.application.reloader.to_prepare do
      # Autoload classes and modules needed at boot time here.
    end

That block runs when the application boots, and every time there is a reload.
For historical reasons, it may run twice, so it has to be idempotent.

Check the "Autoloading and Reloading Constants" guide to learn more about how
Rails autoloads and reloads.
 (called from <main> at /Users/rhymes/Development/forem/config/environment.rb:5)
```

to 

```shell
Unset ENV variable: IMGPROXY_KEY.
Unset ENV variable: IMGPROXY_SALT.
DEPRECATION WARNING: Initialization autoloaded the constants ForemStatsDriver, ForemStatsDrivers, ActsAsForemStatsDriver, and ForemStatsDrivers::DatadogDriver.

Being able to do this is deprecated. Autoloading during initialization is going
to be an error condition in future versions of Rails.

Reloading does not reboot the application, and therefore code executed during
initialization does not run again. So, if you reload ForemStatsDriver, for example,
the expected changes won't be reflected in that stale Class object.

These autoloaded constants have been unloaded.

In order to autoload safely at boot time, please wrap your code in a reloader
callback this way:

    Rails.application.reloader.to_prepare do
      # Autoload classes and modules needed at boot time here.
    end

That block runs when the application boots, and every time there is a reload.
For historical reasons, it may run twice, so it has to be idempotent.

Check the "Autoloading and Reloading Constants" guide to learn more about how
Rails autoloads and reloads.
 (called from <main> at /Users/rhymes/Development/forem/config/environment.rb:5)
```

at startup.

You can still load the app, click around and make sure everything is working.

## Added tests?

- [x] No, and this is why: code has changed and tests need to pass as they are

## [Forem core team only] How will this change be communicated?

_Will this PR introduce a change that impacts Forem members or creators, the
development process, or any of our internal teams? If so, please note how you
will share this change with the people who need to know about it._

      [Admin Guide](https://forem.gitbook.io/forem-admin-guide/), or
      [Storybook](https://storybook.forem.com/) (for Crayons components)
      or in a [forem.dev](http://forem.dev) post
      replace this line with details on why this change doesn't need to be
      shared_
